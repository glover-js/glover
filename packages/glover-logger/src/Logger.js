// @flow
import type {
  LogLevel,
  LogEntry,
  LogType,
  LogMessage,
  LogConfig,
} from 'types/Log';

import addColorForLogType from './addColorForLogType';
import addSymbolForLogType from './addSymbolForLogType';
import LogFile from './LogFile';

export default class Logger {
  _logLevel: LogLevel;
  _logEntries: Array<LogEntry>;
  _useSymbols: boolean;
  _useColors: boolean;
  _saveEntries: boolean;
  _writeToFile: boolean;
  _logFile: LogFile;

  constructor({
    logLevel = 'default',
    useSymbols = true,
    useColors = true,
    saveEntries = false,
    writeToFile = false,
    logFilePath,
  }: LogConfig = {}) {
    this._logLevel = logLevel;
    this._logEntries = [];
    this._useSymbols = useSymbols;
    this._useColors = useColors;
    this._saveEntries = saveEntries;
    this._writeToFile = writeToFile;
    this._logFile = new LogFile(logFilePath);
  }

  log(message: LogMessage): void {
    this._printAndSave(message, 'default');
  }

  info(message: LogMessage): void {
    this._printAndSave(message, 'info');
  }

  success(message: LogMessage): void {
    this._printAndSave(message, 'success');
  }

  warn(message: LogMessage): void {
    this._printAndSave(message, 'warn');
  }

  error(message: LogMessage): void {
    this._printAndSave(message, 'error');
  }

  debug(message: LogMessage): void {
    if (this._logLevel !== 'debug') {
      return;
    }
    this._printAndSave(message, 'debug');
  }

  getLogLevel(): LogLevel {
    return this._logLevel;
  }

  setLogLevel(logLevel: LogLevel): void {
    this._logLevel = logLevel;
  }

  getLogEntries(): Array<LogEntry> {
    return this._logEntries;
  }

  getLogEntriesOfType(type: LogType): Array<LogEntry> {
    return this._logEntries.filter(entry => entry.type === type);
  }

  clearLogEntries(): void {
    this._logEntries = [];
  }

  containsMessage(message: LogMessage): boolean {
    return this._logEntries.findIndex(entry => entry.message === message) > -1;
  }

  _printAndSave(message: LogMessage, type: LogType): void {
    if (this._logLevel !== 'silent') {
      const withSymbol = this._useSymbols
        ? addSymbolForLogType(message, type)
        : message;
      const withColor = this._useColors
        ? addColorForLogType(withSymbol, type)
        : withSymbol;
      console.log(withColor); // eslint-disable-line no-console
    }
    if (this._saveEntries) {
      this._logEntries.push({ message, type });
    }
    if (this._writeToFile) {
      this._logFile.write(`[${type}] ${message}`);
    }
  }
}
