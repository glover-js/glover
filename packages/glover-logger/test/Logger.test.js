import Logger from '../src/Logger';

jest.mock('../src/LogFile.js', () =>
  jest.fn(() => ({
    write: jest.fn(),
  })),
);

function callAllLogMethods(logger) {
  ['log', 'info', 'success', 'warn', 'error', 'debug'].forEach(method => {
    logger[method](`${method}-message`);
  });
}

describe('Tests for Logger', () => {
  describe('logging', () => {
    it('formats and logs messages to the console', () => {
      const logger = new Logger({});
      const log = jest.spyOn(console, 'log');
      log.mockImplementation(() => {});
      callAllLogMethods(logger);
      expect(log.mock.calls).toMatchSnapshot();
      log.mockRestore();
    });
    it('stays silent if logLevel = silent', () => {
      const logger = new Logger({ logLevel: 'silent' });
      const log = jest.spyOn(console, 'log');
      log.mockImplementation(() => {});
      callAllLogMethods(logger);
      expect(log).not.toHaveBeenCalled();
      log.mockRestore();
    });
    it('logs debug messages if logLevel = debug', () => {
      const logger = new Logger({ logLevel: 'debug' });
      const log = jest.spyOn(console, 'log');
      log.mockImplementation(() => {});
      callAllLogMethods(logger);
      expect(log.mock.calls).toMatchSnapshot();
      log.mockRestore();
    });
    it('does not use any symbols if useSymbols = false', () => {
      const logger = new Logger({ logLevel: 'debug', useSymbols: false });
      const log = jest.spyOn(console, 'log');
      log.mockImplementation(() => {});
      callAllLogMethods(logger);
      expect(log.mock.calls).toMatchSnapshot();
      log.mockRestore();
    });
    it('does not use any colors if useColors = false', () => {
      const logger = new Logger({ logLevel: 'debug', useColors: false });
      const log = jest.spyOn(console, 'log');
      log.mockImplementation(() => {});
      callAllLogMethods(logger);
      expect(log.mock.calls).toMatchSnapshot();
      log.mockRestore();
    });
  });
  describe('save entries', () => {
    it('saves all logged messages if saveEntries = true', () => {
      const logger = new Logger({ logLevel: 'silent', saveEntries: true });
      callAllLogMethods(logger);
      expect(logger.getLogEntries()).toMatchSnapshot();
    });
    it('only saves logged messages if saveEntries = true', () => {
      const logger = new Logger({ logLevel: 'silent' });
      callAllLogMethods(logger);
      expect(logger.getLogEntries()).toHaveLength(0);
    });
  });
  describe('write to logfile', () => {
    it('writes all logged messages to a logfile if writeToFile = true', () => {
      const logger = new Logger({ logLevel: 'debug', writeToFile: true });
      const write = jest.spyOn(logger._logFile, 'write'); // eslint-disable-line no-underscore-dangle
      const log = jest.spyOn(console, 'log');
      log.mockImplementation(() => {});
      expect(write.mock.calls).toMatchSnapshot();
      write.mockRestore();
      log.mockRestore();
    });
    it('only writes to logfile if writeToFile = true', () => {
      const logger = new Logger({ logLevel: 'silent' });
      const write = jest.spyOn(logger._logFile, 'write'); // eslint-disable-line no-underscore-dangle
      expect(write.mock.calls).toMatchSnapshot();
      write.mockRestore();
    });
  });
  describe('.getLogEntriesOfType', () => {
    it('returns all log entries of the specified type', () => {
      const logger = new Logger({ logLevel: 'silent', saveEntries: true });
      callAllLogMethods(logger);
      expect(logger.getLogEntriesOfType('warn')).toMatchSnapshot();
    });
  });
  describe('.clearLogEntries', () => {
    it('clears all saved log entries', () => {
      const logger = new Logger({ logLevel: 'silent', saveEntries: true });
      callAllLogMethods(logger);
      expect(logger.getLogEntries()).toHaveLength(5);
      logger.clearLogEntries();
      expect(logger.getLogEntries()).toHaveLength(0);
    });
  });
  describe('.containsMessage', () => {
    it('returns true if the saved log contains the given message', () => {
      const logger = new Logger({ logLevel: 'silent', saveEntries: true });
      logger.log('a message');
      expect(logger.containsMessage('a message')).toEqual(true);
    });
    it('returns false if the saved log does not contain the given message', () => {
      const logger = new Logger({ logLevel: 'silent', saveEntries: true });
      logger.log('a message');
      expect(logger.containsMessage('some other message')).toEqual(false);
    });
  });
  describe('.getLogLevel', () => {
    it('return the log level', () => {
      const silentLogger = new Logger({ logLevel: 'silent' });
      const debugLogger = new Logger({ logLevel: 'debug' });
      const defaultLogger = new Logger();
      expect([
        silentLogger.getLogLevel(),
        debugLogger.getLogLevel(),
        defaultLogger.getLogLevel(),
      ]).toEqual(['silent', 'debug', 'default']);
    });
  });
  describe('.setLogLevel', () => {
    it('sets the log level', () => {
      const logger = new Logger({ logLevel: 'silent' });
      logger.setLogLevel('debug');
      expect(logger.getLogLevel()).toEqual('debug');
    });
  });
});
