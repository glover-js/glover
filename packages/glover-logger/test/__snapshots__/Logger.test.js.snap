// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Logger .getLogEntriesOfType returns all log entries of the specified type 1`] = `
Array [
  Object {
    "message": "warn-message",
    "type": "warn",
  },
]
`;

exports[`Tests for Logger logging does not use any colors if useColors = false 1`] = `
Array [
  Array [
    "   log-message",
  ],
  Array [
    "â„¹  info-message",
  ],
  Array [
    "âœ”  success-message",
  ],
  Array [
    "âš   warn-message",
  ],
  Array [
    "âœ–  error-message",
  ],
  Array [
    "[DEBUG]  debug-message",
  ],
]
`;

exports[`Tests for Logger logging does not use any symbols if useSymbols = false 1`] = `
Array [
  Array [
    "log-message",
  ],
  Array [
    "[34minfo-message[39m",
  ],
  Array [
    "[32msuccess-message[39m",
  ],
  Array [
    "[33mwarn-message[39m",
  ],
  Array [
    "[31merror-message[39m",
  ],
  Array [
    "[35mdebug-message[39m",
  ],
]
`;

exports[`Tests for Logger logging formats and logs messages to the console 1`] = `
Array [
  Array [
    "   log-message",
  ],
  Array [
    "[34mâ„¹  info-message[39m",
  ],
  Array [
    "[32mâœ”  success-message[39m",
  ],
  Array [
    "[33mâš   warn-message[39m",
  ],
  Array [
    "[31mâœ–  error-message[39m",
  ],
]
`;

exports[`Tests for Logger logging logs debug messages if logLevel = debug 1`] = `
Array [
  Array [
    "   log-message",
  ],
  Array [
    "[34mâ„¹  info-message[39m",
  ],
  Array [
    "[32mâœ”  success-message[39m",
  ],
  Array [
    "[33mâš   warn-message[39m",
  ],
  Array [
    "[31mâœ–  error-message[39m",
  ],
  Array [
    "[35m[DEBUG]  debug-message[39m",
  ],
]
`;

exports[`Tests for Logger save entries saves all logged messages if saveEntries = true 1`] = `
Array [
  Object {
    "message": "log-message",
    "type": "default",
  },
  Object {
    "message": "info-message",
    "type": "info",
  },
  Object {
    "message": "success-message",
    "type": "success",
  },
  Object {
    "message": "warn-message",
    "type": "warn",
  },
  Object {
    "message": "error-message",
    "type": "error",
  },
]
`;

exports[`Tests for Logger write to logfile only writes to logfile if writeToFile = true 1`] = `Array []`;

exports[`Tests for Logger write to logfile writes all logged messages to a logfile if writeToFile = true 1`] = `Array []`;
